models/transaction.go

```diff
@@ -1,20 +1,125 @@
 package models
 
 import (
-	"bytes"
 	"encoding/hex"
 	"fmt"
-	"github.com/onflow/flow-go-sdk"
 	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
 	gethTypes "github.com/ethereum/go-ethereum/core/types"
-	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/onflow/cadence"
-	cdcCommon "github.com/onflow/cadence/runtime/common"
 	"github.com/onflow/flow-go/fvm/evm/types"
 )
 
+type Transaction interface {
+	Hash() (common.Hash, error)
+	RawSignatureValues() (v *big.Int, r *big.Int, s *big.Int)
+	From() (common.Address, error)
+	To() *common.Address
+	Data() []byte
+	Nonce() uint64
+	Value() *big.Int
+	Type() uint8
+	Gas() uint64
+	GasPrice() *big.Int
+	BlobGas() uint64
+	Size() uint64
+	MarshalBinary() ([]byte, error)
+}
+
+var _ Transaction = &DirectCall{}
+
+type DirectCall struct {
+	*types.DirectCall
+}
+
+func (dc DirectCall) Hash() (common.Hash, error) {
+	return dc.DirectCall.Hash()
+}
+
+func (dc DirectCall) RawSignatureValues() (
+	v *big.Int,
+	r *big.Int,
+	s *big.Int,
+) {
+	return big.NewInt(0), big.NewInt(0), big.NewInt(0)
+}
+
+func (dc DirectCall) From() (common.Address, error) {
+	return dc.DirectCall.From.ToCommon(), nil
+}
+
+func (dc DirectCall) To() *common.Address {
+	var to *common.Address
+	if !dc.DirectCall.EmptyToField() {
+		ct := dc.DirectCall.To.ToCommon()
+		to = &ct
+	}
+	return to
+}
+
+func (dc DirectCall) Data() []byte {
+	return dc.DirectCall.Data
+}
+
+func (dc DirectCall) Nonce() uint64 {
+	return dc.DirectCall.Nonce
+}
+
+func (dc DirectCall) Value() *big.Int {
+	return dc.DirectCall.Value
+}
+
+func (dc DirectCall) Type() uint8 {
+	return dc.DirectCall.Type
+}
+
+func (dc DirectCall) Gas() uint64 {
+	return dc.DirectCall.GasLimit
+}
+
+func (dc DirectCall) GasPrice() *big.Int {
+	return big.NewInt(0)
+}
+
+func (dc DirectCall) BlobGas() uint64 {
+	return 0
+}
+
+func (dc DirectCall) Size() uint64 {
+	encoded, err := dc.MarshalBinary()
+	if err != nil {
+		return 0
+	}
+	return uint64(len(encoded))
+}
+
+func (dc DirectCall) MarshalBinary() ([]byte, error) {
+	return dc.DirectCall.Encode()
+}
```

Comment:
why not simplify like I suggested here: https://github.com/onflow/flow-evm-gateway/pull/132#discussion_r1516216917

Reply:
To get this to work, it needs some more method overrides, e.g.:
```go
type DirectCall struct {
	*gethTypes.Transaction
	call *types.DirectCall
}

func (dc DirectCall) Hash() (common.Hash, error) {
	return dc.call.Hash()
}

func (dc DirectCall) From() (common.Address, error) {
	return dc.call.From.ToCommon(), nil
}

func (dc DirectCall) Type() uint8 {
	return dc.call.Type
}

func (dc DirectCall) Size() uint64 {
	encoded, err := dc.MarshalBinary()
	if err != nil {
		return 0
	}
	return uint64(len(encoded))
}

func (dc DirectCall) MarshalBinary() ([]byte, error) {
	return dc.call.Encode()
}
```

And then the instantiation becomes:

```go
DirectCall{Transaction: directCall.Transaction(), call: directCall}
```

Although it requires fewer method implementations, the resulting interface of `models.DirectCall` is more bloated from `*gethTypes.Transaction`. Personally, I find the end-result to be more complicated. But if you have a strong opinion about it, I can change it.

Reply:
Hmm I'm not sure I understand this: 
> the resulting interface of models.DirectCall is more bloated from *gethTypes.Transaction

I would say having less code is better in this case, I also believe `Hash()` can be dropped as it's just the same. But I want to understand your concerns before say to use my suggestion.

Reply:
> I also believe Hash() can be dropped as it's just the same

Not yet, because we still have `Hash() (common.Hash, error)`. When we drop the `error` return value from flow-go, then we can drop it I guess.

> Hmm I'm not sure I understand this:

I mean that now, on objects of `models.DirectCall` there are available all the methods of `*gethTypes.Transaction`. These might be leaking implementation details, and they do not necessarily make sense for a `DirectCall`. Someone might see the `DirectCall{}.BlobHashes()` method, and mistakenly assume that it is a regular geth transaction. That's what I mean by bloated interface, the set of available method from `*gethTypes.Transaction`.